### DISCUSSION TITLE
How to generate class variables depending on command line?

### DISCUSSION THREAD

**Discussion by GeJintian** [pytest-dev/pytest#9101] [Q&A] [answered]
Hello everyone, I have a small question about pytest command line parameters. In my case, I want to open a .txt file, read some information from it and run some tests on these information. Since the name of the .txt file is not fixed, I want to pass it by command line. Also, I try to enclose all my test functions in a class, so that I only need to read the file once. Here is how I plan to implement this:<br>
```python
@pytest.fixture(scope='class')
def setup_class(tester):
    print("enter setup")
    dir ='dataset/lidar'
    tester=tester.split('.')[0]
    with open(dir+'/result/'+tester+'.pcd.json','r',encoding='utf8')as f:
        test=json.load(f)
    with open(dir+'/gt/'+tester+'.json','r',encoding='utf8')as f:
        standard=json.load(f)
    return test, standard

class TestSimple:
    def test_rotate_x(self,setup_class):
        test,standard=setup_class
        assert abs(test['extrinsic_param']['rotation'][0]-standard['extrinsic_param']['rotation'][0]) < 0.01
```
and in conftest.py, I write:<br>
```python
def pytest_addoption(parser):
    parser.addoption(
        "--tester", action="store", default="lidar", help="my option: type1 or type2"
    )
@pytest.fixture
def tester(request):
    return request.config.getoption("--tester")
```
But I got an error like
```bash
________________________________________________ ERROR at setup of TestSimple.test_rotate_x _________________________________________________
ScopeMismatch: You tried to access the 'function' scoped fixture 'tester' with a 'class' scoped request object, involved factories
test_front_lidar.py:8:  def setup_class(tester)
conftest.py:8:  def tester(request)
```
Does anyone have used a command line parameter to initialize class variables? Or is there a better way to achieve this goal? Thank you very much to have a look.

**Comment by GeJintian** [comment, marked as answer]
OK, it seems like adding one line suffice to solve my problem:<br>
```python
def setup_class(request):
    tester=request.config.getoption("--tester")
    dir ='/dataset/lidar'
    tester=tester.split('.')[0]
    with open(dir+'/result/'+tester+'.pcd.json','r',encoding='utf8')as f:
        test=json.load(f)
    with open(dir+'/gt/'+tester+'.json','r',encoding='utf8')as f:
        standard=json.load(f)
    return test, standard
```

**Reply by RonnyPfannschmidt** [reply]
@GeJintian setting the scope of the tester fixture would also have solved it

**Reply by GeJintian** [reply]
@RonnyPfannschmidt  Thank you very much for your reply. Meanwhile, I wonder that if there is a method by which I can repeat a Test class over several datasets? For example, I have 10 data files with their names stored in a .txt file. I want to apply my Test class on all of these files and generate only 1 summary report (in this case there should be 10 tests in that report). Do you have any suggestions about this? Thank you very much.

**Reply by RonnyPfannschmidt** [reply]
@GeJintian if you alter the tester fixture to return `request.param` if its existing and the option otherwise,
you can as a next step start to experiment with pytest.mark.parametrize to pass in the tester values via a indirect key

another possible option is to have a pytest plugin that declares the fixture at pytest-configure time and adds the params parameter nicely
